File: src/main/java/aoc2015/day01/Day01.java
package aoc2015.day01;

import aoc2015.AbstractDay2015;
import lombok.NonNull;

public class Day01 extends AbstractDay2015 {

  Day01() {
    super(1);
  }

  @NonNull
  @Override
  public Integer part1(@NonNull String input) {
    int floor = 0;

    for (char c : input.toCharArray()) {
      if (c == '(') {
        floor += 1;
      } else {
        floor -= 1;
      }
    }

    return floor;
  }

  @NonNull
  @Override
  public Integer part2(@NonNull String input) {
    int floor = 0;

    for (int i = 0; i < input.length(); i++) {
      char c = input.charAt(i);

      if (c == '(') {
        floor += 1;
      } else {
        floor -= 1;
      }

      if (floor == -1) {
        return i + 1;
      }
    }

    return -1;
  }

  public static void main(String[] args) {
    new Day01().executeParts();
  }
}

File: aoc25/src/main/java/day01/Day01.kt
package day01

import AbstractDay2025
import kotlin.math.abs


fun main() {
    Day01().executeParts()
}

class Day01 : AbstractDay2025(1) {

    private val radius = 100

    override fun part1(input: String): Any {
        val rotations = parseInput(input)
        var rotation = 50
        var timesAtZero = 0

        for (i in rotations) {
            rotation = (rotation + i).mod(radius)
            if (rotation == 0) {
                timesAtZero++
            }
        }

        return timesAtZero
    }

    override fun part2(input: String): Any {
        val rotations = parseInput(input)
        var rotation = 50
        var timesPassingZero = 0

        for (i in rotations) {
            if (i >= 0) {
                timesPassingZero += (rotation + i) / radius
            } else {
                timesPassingZero += ((radius - rotation) % radius + abs(i)) / radius
            }

            rotation = (rotation + i).mod(radius)
        }

        return timesPassingZero
    }

    fun parseInput(input: String): List<Int> = input.lines()
        .stream()
        .map { parseLine(it) }
        .toList()

    fun parseLine(line: String): Int {
        val rotation = line[0]
        val number = line.substring(1)

        if (rotation == 'L') {
            return -Integer.parseInt(number)
        }
        if (rotation == 'R') {
            return Integer.parseInt(number)
        }
        throw IllegalArgumentException(number)
    }
}